demo/src/components/Cursors.tsx

```diff
@@ -1,127 +1,91 @@
-    position?: {
-      x: number;
-      y: number;
-    };
-  };
-}
 
-interface State {
-  [connectionId: string]: Member & Action['data'];
-}
-
-const reducer = (state: State, action: Action): State => {
-  const { type, data } = action;
-  const { connectionId, members, position } = data;
-  switch (type) {
-    case CURSOR_ENTER:
-      return {
-        ...state,
-        [connectionId]: {
-          ...assign(find(members, { connectionId }), { connectionId, position }),
-        },
-      };
-    case CURSOR_LEAVE:
-      return {
-        ...omit(state, connectionId),
-      };
-    case CURSOR_MOVE:
-      return {
-        ...state,
-        [connectionId]: {
-          ...assign(find(members, { connectionId }), { connectionId, position }),
-        },
-      };
-    default:
-      throw new Error('Unknown dispatch type');
-  }
-};
+type state = typeof CURSOR_ENTER | typeof CURSOR_LEAVE | typeof CURSOR_MOVE;
+type CursorUpdate = Omit<_CursorUpdate, 'data'> & { data: { state: state } };
 
 export const Cursors = () => {
   const space = useContext(SpacesContext);
-  const { self, members } = useMembers();
-  const [activeCursors, dispatch] = useReducer(reducer, {});
+  const { self, others } = useMembers();
+  const [cursors, setCursors] = useState<{
+    [connectionId: string]: { position: CursorUpdate['position']; state: CursorUpdate['data']['state'] };
+  }>({});
 
   useEffect(() => {
-    if (!space || !members) return;
+    if (!space || !others) return;
 
     space.cursors.subscribe('update', (cursorUpdate) => {
-      const { connectionId } = cursorUpdate;
-      const member = find<Member>(members, { connectionId });
+      const { connectionId, position, data } = cursorUpdate as CursorUpdate;
 
-      if (
-        connectionId !== self?.connectionId &&
-        member?.location?.slide === self?.location?.slide &&
-        cursorUpdate.data
-      ) {
-        dispatch({
-          type: cursorUpdate.data.state as ActionType,
-          data: {
-            connectionId,
-            members,
-            position: cursorUpdate.position,
-          },
-        });
-      } else {
-        dispatch({
-          type: CURSOR_LEAVE,
-          data: {
-            connectionId,
-            members,
-          },
-        });
-      }
+      if (cursorUpdate.connectionId === self?.connectionId) return;
+
+      setCursors((currentCursors) => ({
+        ...currentCursors,
+        [connectionId]: { position, state: data.state },
+      }));
     });
 
     return () => {
       space.cursors.unsubscribe('update');
     };
-  }, [space, members]);
+  }, [space, others, self?.connectionId]);
+
+  useEffect(() => {
+    space?.members.subscribe('leave', async (member) => {
+      setCursors((currentCursors) => ({
+        ...currentCursors,
+        [member.connectionId]: { position: { x: 0, y: 0 }, state: CURSOR_LEAVE },
+      }));
+    });
+
+    return () => {
+      space?.members.unsubscribe('leave');
```

Comment:
As written, doesn't this remove _all_ listeners for the `leave` event? Don't we only want to remove the specific listener that we added above?

Reply:
Good catch, yes we should